File: packages/typra-emitter/fixtures/integration/vector-adapters/csharp/VectorAdapters.cs
// Reference adapters for the integration fixture's @vector suite. Copied into
// the C# target's tests/ dir by validate-fixtures before the generated
// conformance suite runs. See fixtures/integration/vector-adapters.
#nullable enable
using System;
using System.Collections.Generic;
using System.Text.Json.Nodes;

namespace Typra.Fixtures.Conformance;

public sealed class VectorContext
{
    public string Contract { get; init; } = "";
    public string Operation { get; init; } = "";
    public JsonNode? Vector { get; init; }
    public string? Provider { get; init; }
    public string? TargetApi { get; init; }
    public JsonNode? Doubles { get; init; }
    public string BaseDir { get; init; } = "";
}

public sealed class VectorException : Exception
{
    public JsonNode? Payload { get; }
    public VectorException(string message, JsonNode? payload = null)
        : base(message) { Payload = payload; }
}

public sealed class VectorAdapter
{
    public required Func<JsonNode?, VectorContext, JsonNode?> Invoke { get; init; }
    public Func<JsonNode?, VectorContext, JsonNode?>? Normalize { get; init; }
}

public static class VectorAdapters
{
    private static JsonNode? AuthorizeInvoke(JsonNode? input, VectorContext ctx) =>
        new JsonObject { ["approved"] = true };

    private static JsonNode? FormatInvoke(JsonNode? input, VectorContext ctx) =>
        input?["messages"]?.DeepClone();

    public static IReadOnlyDictionary<string, VectorAdapter> Adapters() =>
        new Dictionary<string, VectorAdapter>
        {
            ["CanonicalEnginePort.authorize"] = new VectorAdapter { Invoke = AuthorizeInvoke },
            ["CanonicalEnginePort.format"] = new VectorAdapter { Invoke = FormatInvoke },
        };

    public static IReadOnlyDictionary<string, string> Waivers() =>
        new Dictionary<string, string>();

    public static JsonNode? Doubles() => new JsonObject();
}

File: packages/typra-emitter/fixtures/integration/vector-adapters/rust/vector_adapters.rs
// Reference adapters for the integration fixture's @vector suite. Included by
// the generated conformance suite via #[path]. Copied into each Rust target's
// tests/ dir by validate-fixtures. See fixtures/integration/vector-adapters.
#![allow(dead_code, unused_variables, clippy::all)]

use serde_json::{json, Value};
use std::collections::HashMap;

pub struct Context {
    pub contract: String,
    pub operation: String,
    pub vector: Value,
    pub provider: Option<String>,
    pub target_api: Option<String>,
    pub doubles: Value,
    pub base_dir: String,
}

pub struct VectorError {
    pub message: String,
    pub payload: Option<Value>,
}

pub struct Adapter {
    pub invoke: fn(&Value, &Context) -> Result<Value, VectorError>,
    pub normalize: Option<fn(&Value, &Context) -> Value>,
}

fn authorize_invoke(_input: &Value, _ctx: &Context) -> Result<Value, VectorError> {
    Ok(json!({ "approved": true }))
}

fn format_invoke(input: &Value, _ctx: &Context) -> Result<Value, VectorError> {
    Ok(input.get("messages").cloned().unwrap_or(Value::Null))
}

pub fn doubles() -> Value {
    json!({})
}

pub fn waivers() -> HashMap<&'static str, &'static str> {
    HashMap::new()
}

pub fn adapters() -> HashMap<&'static str, Adapter> {
    let mut map = HashMap::new();
    map.insert(
        "CanonicalEnginePort.authorize",
        Adapter { invoke: authorize_invoke, normalize: None },
    );
    map.insert(
        "CanonicalEnginePort.format",
        Adapter { invoke: format_invoke, normalize: None },
    );
    map
}
